src/commands/utils/uploadUDF.test.ts

```diff
@@ -271,24 +283,164 @@ describe("uploadUDF", () => {
+
+      const errorNotificationStub = sandbox
+        .stub(notifications, "showErrorNotificationWithButtons")
+        .resolves();
+
+      await assert.rejects(
+        () => uploadArtifactToCCloud(mockParams, mockUploadId),
+        (err: Error) => err.message === "Upload failed",
+      );
+
+      sinon.assert.calledOnce(errorNotificationStub);
+      sinon.assert.calledWith(
+        errorNotificationStub,
+        "Failed to create Flink artifact. See logs for details.",
+      );
+    });
+    it("should parse and display JSON error details from ResponseError", async () => {
+      const mockUploadId = "upload-id-123";
+      let stubbedFlinkArtifactsApi = sandbox.createStubInstance(FlinkArtifactsArtifactV1Api);
+      const jsonBody = { error: "artifact already exists" };
+      const mockHeaders = new Map([["content-type", "application/json"]]);
+      const mockResponse = {
+        headers: {
+          get: (key: string) => mockHeaders.get(key),
+        },
+        status: 409,
+        json: sandbox.stub().resolves(jsonBody),
+        text: sandbox.stub(),
+      };
+      const responseError = new ResponseError(409, "Conflict", mockResponse as any);
+
+      stubbedFlinkArtifactsApi.createArtifactV1FlinkArtifact.rejects(responseError);
+      let stubbedSidecarHandle = getSidecarStub(sandbox);
+      stubbedSidecarHandle.getFlinkArtifactsApi.returns(stubbedFlinkArtifactsApi);
+
+      const errorNotificationStub = sandbox
+        .stub(notifications, "showErrorNotificationWithButtons")
+        .resolves();
+
+      await assert.rejects(
+        () => uploadArtifactToCCloud(mockParams, mockUploadId),
+        (err: Error) => err === responseError,
+      );
+
+      sinon.assert.calledOnce(errorNotificationStub);
+      sinon.assert.calledWith(
+        errorNotificationStub,
+        `Failed to create Flink artifact: ${JSON.stringify(jsonBody)}`,
+      );
+    });
+
+    it("should parse and display text error details from ResponseError", async () => {
+      const mockUploadId = "upload-id-123";
+      let stubbedFlinkArtifactsApi = sandbox.createStubInstance(FlinkArtifactsArtifactV1Api);
+
+      const textBody = "artifact already exists";
+      const mockHeaders = new Map([["content-type", "text/plain"]]);
+      const mockResponse = {
+        headers: {
+          get: (key: string) => mockHeaders.get(key),
+        },
+        status: 409,
+        json: sandbox.stub(),
+        text: sandbox.stub().resolves(textBody),
+      };
+      const responseError = new ResponseError(409, "Conflict", mockResponse as any);
+
+      stubbedFlinkArtifactsApi.createArtifactV1FlinkArtifact.rejects(responseError);
+      let stubbedSidecarHandle = getSidecarStub(sandbox);
+      stubbedSidecarHandle.getFlinkArtifactsApi.returns(stubbedFlinkArtifactsApi);
+
+      const errorNotificationStub = sandbox
+        .stub(notifications, "showErrorNotificationWithButtons")
+        .resolves();
+
+      await assert.rejects(
```

Comment:
assuming this is the way to coerce the stub into throwing the right error, like in resourceManager:
```
    await assert.rejects(
      rm.setEnvironments(CCLOUD_CONNECTION_ID, mixedEnvironments),
      (err) => {
        return err instanceof Error && err.message.includes("Connection ID mismatch");
      },
      "Expected error when setting mixed connection IDs",
    );
```

but first time doing it so feedback appreciate

Reply:
The `stubbedFlinkArtifactsApi.createArtifactV1FlinkArtifact.rejects(responseError);` is the setup that is going to get the codepath to encounter a raised Error.

The `await assert.rejects(...)` is the way to prove that the tested function's returned Promise inidicates an error was raised that matches some criteria. If no error were raised (that matched the error predicate), then `await assert.rejects` will indeed make the test fail.